app/org/sagebionetworks/bridge/services/ConsentServiceImpl.java

```diff
@@ -168,19 +175,74 @@ public boolean hasUserSignedMostRecentConsent(StudyIdentifier studyIdentifier, U
     }
 
     @Override
-    public void withdrawConsent(Study study, User user) {
+    public void withdrawConsent(Study study, User user, Withdrawal withdrawal) {
         checkNotNull(study, Validate.CANNOT_BE_NULL, "study");
         checkNotNull(user, Validate.CANNOT_BE_NULL, "user");
-        if (userConsentDao.withdrawConsent(user.getHealthCode(), study)) {
-            decrementStudyEnrollment(study);
-            Account account = accountDao.getAccount(study, user.getEmail());
-            account.setConsentSignature(null);
+        checkNotNull(user, Validate.CANNOT_BE_NULL, "withdrawal");
+        
+        String externalId = optionsService.getOption(user.getHealthCode(), ParticipantOption.EXTERNAL_IDENTIFIER);
+        
+        Account account = accountDao.getAccount(study, user.getEmail());
+        signatureToHistory(account);
+        accountDao.updateAccount(study, account);
+
+        try {
+            userConsentDao.withdrawConsent(user.getHealthCode(), study, withdrawal.getWithdrewOn());    
+        } catch(Exception e) {
+            // Could not record the consent, compensate and rethrow the exception
+            signatureFromHistory(account);
```

Comment:
Instead of using a compensating transaction, I feel it would be simpler to just have consentSignatureHistory be canonical. The history would always contain all consent signatures, including the current active signature.

If we're worried about migration, we could just read through to the old consentSignature when consentSignatureHistory was missing or empty, but for any new consents, we write all the consent signatures to consent signature history. At this point, there's no rush to do the migration, and we can launch the feature before migration.